fdbserver/KeyValueStoreRocksDB.actor.cpp

```diff
@@ -312,6 +314,266 @@ class ReadIteratorPool {
+			{ "MergeOperatorTimeNanos", rocksdb_merge_operator_time_nanos, {} },
+			{ "ReadIndexBlockNanos", rocksdb_read_index_block_nanos, {} },
+			{ "ReadFilterBlockNanos", rocksdb_read_filter_block_nanos, {} },
+			{ "NewTableBlockIterNanos", rocksdb_new_table_block_iter_nanos, {} },
+			{ "NewTableIteratorNanos", rocksdb_new_table_iterator_nanos, {} },
+			{ "BlockSeekNanos", rocksdb_block_seek_nanos, {} },
+			{ "FindTableNanos", rocksdb_find_table_nanos, {} },
+			{ "BloomMemtableHitCount", rocksdb_bloom_memtable_hit_count, {} },
+			{ "BloomMemtableMissCount", rocksdb_bloom_memtable_miss_count, {} },
+			{ "BloomSstHitCount", rocksdb_bloom_sst_hit_count, {} },
+			{ "BloomSstMissCount", rocksdb_bloom_sst_miss_count, {} },
+			{ "KeyLockWaitTime", rocksdb_key_lock_wait_time, {} },
+			{ "KeyLockWaitCount", rocksdb_key_lock_wait_count, {} },
+			{ "EnvNewSequentialFileNanos", rocksdb_env_new_sequential_file_nanos, {} },
+			{ "EnvNewRandomAccessFileNanos", rocksdb_env_new_random_access_file_nanos, {} },
+			{ "EnvNewWritableFileNanos", rocksdb_env_new_writable_file_nanos, {} },
+			{ "EnvReuseWritableFileNanos", rocksdb_env_reuse_writable_file_nanos, {} },
+			{ "EnvNewRandomRwFileNanos", rocksdb_env_new_random_rw_file_nanos, {} },
+			{ "EnvNewDirectoryNanos", rocksdb_env_new_directory_nanos, {} },
+			{ "EnvFileExistsNanos", rocksdb_env_file_exists_nanos, {} },
+			{ "EnvGetChildrenNanos", rocksdb_env_get_children_nanos, {} },
+			{ "EnvGetChildrenFileAttributesNanos", rocksdb_env_get_children_file_attributes_nanos, {} },
+			{ "EnvDeleteFileNanos", rocksdb_env_delete_file_nanos, {} },
+			{ "EnvCreateDirNanos", rocksdb_env_create_dir_nanos, {} },
+			{ "EnvCreateDirIfMissingNanos", rocksdb_env_create_dir_if_missing_nanos, {} },
+			{ "EnvDeleteDirNanos", rocksdb_env_delete_dir_nanos, {} },
+			{ "EnvGetFileSizeNanos", rocksdb_env_get_file_size_nanos, {} },
+			{ "EnvGetFileModificationTimeNanos", rocksdb_env_get_file_modification_time_nanos, {} },
+			{ "EnvRenameFileNanos", rocksdb_env_rename_file_nanos, {} },
+			{ "EnvLinkFileNanos", rocksdb_env_link_file_nanos, {} },
+			{ "EnvLockFileNanos", rocksdb_env_lock_file_nanos, {} },
+			{ "EnvUnlockFileNanos", rocksdb_env_unlock_file_nanos, {} },
+			{ "EnvNewLoggerNanos", rocksdb_env_new_logger_nanos, {} },
+		};
+		for (auto& t : metrics) { // readers, then writer
+			auto& [name, metric, vals] = t;
+			for (int i = 0; i < SERVER_KNOBS->ROCKSDB_READ_PARALLELISM; i++) {
+				vals.push_back(0); // add reader
+			}
+			vals.push_back(0); // add writer
+		}
+	}
+	void reset() { rocksdb::get_perf_context()->Reset(); }
+	void setWriter() {
+		for (auto& t : metrics) {
+			auto& [name, metric, vals] = t;
+			vals[SERVER_KNOBS->ROCKSDB_READ_PARALLELISM] = getRocksdbPerfcontextMetric(metric);
+		}
+	}
+	void setReader(int rdidx) {
+		for (auto& t : metrics) {
+			auto& [name, metric, vals] = t;
+			vals[rdidx] = getRocksdbPerfcontextMetric(metric);
```

Comment:
Why can't we get the report? calling each metric would be sightly expensive right?

Reply:
No, the overhead is potentially high. When using report, we need do `perf=rocksdb_perfcontext_create()`, then do `rocksdb_perfcontext_report(perf, true)` for each db operation. Note that `rocksdb_perfcontext_create()` news a new object for each call. Since perf->PerfContext is threadLocal variable, we cannot do `rocksdb_perfcontext_create()` once when initialization and gets reports in db operations later, because the laters operations may not in the same thread as the initialization thread.